iocore/net/P_SSLConfig.h

```diff
@@ -42,6 +49,211 @@
+  {
+    _size = that._size;
+    if (PERMANENT == that._duration) {
+      _duration = PERMANENT;
+      _ptr      = that._ptr;
+    } else {
+      _duration = OWNED_IN_HEAP;
+      char *p   = new char[sizeof(T) * _size];
+      for (std::size_t i = 0; i < _size; ++i, p += sizeof(T)) {
+        ::new (p) T(that._ptr[i]);
+      }
+      _ptr = static_cast<T *>(p);
+    }
+  }
+
+  DurableConstArray(DurableConstArray &&that)
+  {
+    _duration = that._duration;
+    _ptr      = that._ptr;
+    _size     = that._size;
+
+    ::new (&that) DurableConstArray();
+  }
+
+  DurableConstArray &
+  operator=(const DurableConstArray &that)
+  {
+    if (this != &that) {
+      this->~DurableConstArray();
+
+      ::new (this) DurableConstArray(that);
+    }
+    return *this;
+  }
+
+  DurableConstArray &
+  operator=(DurableConstArray &&that)
+  {
+    this->~DurableConstArray();
+
+    _duration = that._duration;
+    _ptr      = that._ptr;
+    _size     = that._size;
+
+    ::new (&that) DurableConstArray();
+
+    return *this;
+  }
+
+  Duration
+  duration() const
+  {
+    return _duration;
+  }
+  T const *
+  data() const
+  {
+    return _ptr;
+  }
+  std::size_t
+  size() const
+  {
+    return _size;
+  }
+
+  friend typename std::enable_if<THasEq, bool>::type
+  operator==(DurableConstArray const &a, DurableConstArray const &b)
+  {
+    if (a._size != b._size) {
+      return false;
+    }
+    if (a._ptr == b._ptr) {
+      return true;
+    }
+    if (!(a._ptr && b._ptr)) {
+      return false;
+    }
+    for (std::size_t i = 0; i < a._size; ++i) {
+      if (a._ptr[i] != b._ptr[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  friend typename std::enable_if<THasEq, bool>::type
+  operator!=(DurableConstArray const &a, DurableConstArray const &b)
+  {
+    return !(a == b);
+  }
+
+private:
+  Duration _duration{PERMANENT};
+  T const *_ptr{nullptr};
+  std::size_t _size{0};
+};
+
+class TwoCStrKey
+{
+public:
+  TwoCStrKey() {}
+
+  TwoCStrKey(char const *major, char const *minor) : _major(_make(major)), _minor(_make(minor)) {}
+
+  friend bool
+  operator==(TwoCStrKey const &a, TwoCStrKey const &b)
+  {
+    if (a._major != b._major) {
+      return false;
+    }
+    return a._minor == b._minor;
+  }
+
+  std::string_view
+  major() const
+  {
+    return std::string_view(_major.data(), _major.size());
+  }
+  std::string_view
+  minor() const
+  {
+    return std::string_view(_minor.data(), _minor.size());
+  }
+
+  friend struct TwoCStrKeyHash;
+
+private:
+  using Comp = DurableConstArray<char, true>;
+
+  Comp _major, _minor;
+
+  Comp
+  _make(char const *cStr)
+  {
+    if (!cStr || !*cStr) {
+      return Comp();
+    }
+    return Comp::make(Comp::UNOWNED_TRANSIENT, cStr, std::strlen(cStr));
+  }
+};
+
+const std::uint32_t FnvHashInit = 0x811c9dc5;
+
+// FNV-1a hash
+inline std::uint32_t
+fnvHashNext(std::uint8_t next, std::uint32_t hash = FnvHashInit)
```

Comment:
Can we use "include/tscore/HashFNV.h"? I don't want to have two implementations.

Reply:
Yes it would only be slightly slower.  Fortunately gcc is able to optimize the optimization:  http://coliru.stacked-crooked.com/a/85fab3da8be52c13